docs/generatedcodeguide.rst

```diff
@@ -0,0 +1,228 @@
+    record messageName {
+    ...
+    }
+  }
+
+The same proto file with the a package declaration ``package myPackage;`` will
+result in a file called ``myPackage.chpl``.
+
+.. code-block:: chpl
+
+  module myPackage {
+    record messageName {
+    ...
+    }
+  }
+
+Messages
+--------
+Given a simple message declaration:
+
+.. code-block:: proto
+  
+  message Foo {
+    int32 num = 1;
+  }
+  
+The protocol buffer compiler generates a record called ``Foo``, which have message
+field initializers and serialization/parsing methods for wire-type encoding.
+
+ .. code-block:: chpl
+  
+  record Foo {
+
+    // Record fields will be generated corresponding to each proto message field.
+    var num: int(32);
+    
+    // Used to store encoded byte stream of unknown fields encountered while parsing.
+    // As per proto3 documentation, unknown fields should be preserved and appended
+    // to the generated message byte stream.
+    var unknownFieldStream: bytes = "";
+    
+    // User exposed method for serializing data to protobuf wire format.
+    // This is a wrapper method to the actual method, it is used for making a
+    // binary writing channel out of the channel given as an input by the user. 
+    proc serialize(ch) throws { ... }
+    
+    // Contains the actual implementation for serializing data. Calls the
+    // append functions of the user support library. It appends the `unknownFieldStream`
+    // at the end of the message.
+    proc _serialize(binCh) throws { ... }
+    
+    // User exposed method for parsing data from protobuf wire format.
+    // This is a wrapper method to the actual method, it is used for making a
+    // binary reading channel out of the channel given as an input by the user. 
+    proc deserialize(ch) throws { ... }
+    
+    // Contains the actual implementation for parsing data. Calls the consume
+    // functions of the user support library. Appends unknown fields encountered
+    // to the `unknownFieldStream` variable.
+    proc _deserialize(binCh) throws { ... }
+  
+  }  
+
+Nested Types
+------------
+A message can be declared inside another message. For example:
+
+.. code-block:: proto
+
+  message Foo {
+    message Bar {
+      ...
+    }
+  }
+
+In this case, or if a message contains a nested enum, the compiler will generate
+module level records/enums with a name prefixed by the parent message name:
+
+.. code-block:: chpl
+  
+  record Foo {
+   ...
+  }
+  
+  // Nested Types
+  record Foo_Bar {
+    ...
+  }
+
+.. note::
+  Nested records or declaration of enums in records is currently not supported in
+  Chapel. Once we have support for these, we can declare nested type in the parent
+  record and thus avoiding the name prefix. 
+
+Fields
+------
+The protocol buffer compiler generates a Chapel record field for each field defined
+within a message. In case of Chapel, methods equivalent to `get` and `set` in other
+languages are implicitly generated by the Chapel compiler.
```

Comment:
I might remove "In case of Chapel" from this sentence and just start with "Methods equivalent".  Then I'd finish the sentence with ", so do not need to be generated by the protocol buffer compiler."